moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp

```diff
@@ -79,14 +79,31 @@ bool GlobalPlannerComponent::initializeGlobalPlanner()
       },
       [this](const std::shared_ptr<rclcpp_action::ServerGoalHandle<moveit_msgs::action::GlobalPlanner>>& /*unused*/) {
         RCLCPP_INFO(LOGGER, "Received request to cancel global planning goal");
+        if (long_callback_thread_ != nullptr)
+        {
+          long_callback_thread_->join();
+          long_callback_thread_ = nullptr;
+        }
         if (!global_planner_instance_->reset())
         {
           throw std::runtime_error("Failed to reset the global planner while aborting current global planning");
         }
         return rclcpp_action::CancelResponse::ACCEPT;
       },
       [this](std::shared_ptr<rclcpp_action::ServerGoalHandle<moveit_msgs::action::GlobalPlanner>> goal_handle) {
-        return globalPlanningRequestCallback(goal_handle);
+        // this needs to return quickly to avoid blocking the executor, so spin up a new thread
+        if (long_callback_thread_ == nullptr)
+        {
+          long_callback_thread_ =
+              std::make_unique<std::thread>(&GlobalPlannerComponent::globalPlanningRequestCallback, this, goal_handle);
+        }
+        else
+        {
+          long_callback_thread_->join();
```

Comment:
I don't want to make this more complicated than it already is. I wish `join` had a timeout but it doesn't. I'm sure @tylerjw has a strong opinion   ;)

Reply:
I think @zflat is right, that this wouldn't solve thread synchronization. Indeed, rejecting the goal would be a valid solution since we only support one thread at a time. Also, it wouldn't have to be joined in that case, the action would have to be canceled first. For this we wouldn't even need a lock guard, you could simply check if the thread is active (=joinable).

Reply:
> rejecting the goal would be a valid solution since we only support one thread at a time

Well, that doesn't really make sense with the way RE2 is using this. We need to be able to accept new goals in a rapid-fire way.

Reply:
How about I put a timeout around `join` like this. At least we will be able to throw a nice, human-readable error if thread joining takes too long.  (From my testing, it doesn't seem to be a problem in reality.)

https://stackoverflow.com/a/33017819

```
// Terminate the thread.
auto future = std::async(std::launch::async, &std::thread::join, &thread);
if (future.wait_for(std::chrono::seconds(5)) 
    == std::future_status::timeout) {
  /* --- Do something, if thread has not terminated within 5 s. --- */
}
```

Reply:
https://user-images.githubusercontent.com/11284393/171268837-c72669b7-b1e0-4a30-8cef-2fedf99ced97.mp4